### PR TITLE
Introduce idx_prev to use as iguess

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes half of https://github.com/SciML/DataInterpolations.jl/issues/271

## Checklist

- [x] Appropriate tests were added (n/a)
- [x] Any code changes were done in a way that does not break public API
- [x] All documentation related to code changes were updated
- [x] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [x] Any new documentation only uses public API

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
Quick benchmark:

```julia
using DataInterpolations
using BenchmarkTools

N = 1000
u = rand(N)
t = cumsum(rand(N))

itp = LinearInterpolation(u, t)

t_eval = range(first(t), last(t), length = 10000)

@btime itp.(t_eval)

# old: 276.000 μs (3 allocations: 78.25 KiB)
# new: 166.600 μs (3 allocations: 78.27 KiB)
```

So the better `iguess` gives a speedup because successive evaluations are close together. What kind of benchmarking would you like to see?

**Review by ChrisRackauckas — DISMISSED** [review]

**Review by sathvikbhagavan — COMMENTED** [review]
Looks good! Can you add some tests to check whether `idx_prev` is properly set?

File: test/derivative_tests.jl [hunk 1]
```diff
@@ -35,6 +36,11 @@ function test_derivatives(method, u, t; args = [], kwargs = [], name::String)
             adiff2 = derivative(func, _t, 2)
             @test isapprox(fdiff, adiff, atol = 1e-8)
             @test isapprox(fdiff2, adiff2, atol = 1e-8)
+            # Cached index
+            if hasproperty(func, :idx_prev)
+                @test abs(func.idx_prev[] -
```
Comment (sathvikbhagavan) [inline test/derivative_tests.jl:41]: Can this test be written with equality?

File: test/interpolation_tests.jl [hunk 2]
```diff
@@ -1,7 +1,18 @@
 using DataInterpolations
+using FindFirstFunctions: searchsortedfirstcorrelated
 using StableRNGs
 using Optim, ForwardDiff
 
+function test_cached_index(A)
+    for t in range(first(A.t), last(A.t); length = 2 * length(A.t) - 1)
+        A(t)
+        if hasproperty(A, :idx_prev)
+            @test abs(A.idx_prev[] -
```
[the remark this answers is not in the document]
Comment (SouthEndMusic) [reply test/interpolation_tests.jl]: As the kwargs of the new function `get_idx` show, the index is obtained in several slightly different ways. Therefore it is not possible to test the index with equality in an interpolation method agnostic way :/

File: src/interpolation_utils.jl [hunk 3]
```diff
@@ -54,11 +55,21 @@ function munge_data(U::StridedMatrix, t::AbstractVector)
...
+    newt = Tt.([t[idx] for idx in non_missing_indices])
 
     return hcat(newUs...), newt
 end
+
+function get_idx(tvec, t, iguess; lb = 1, ub_shift = -1, idx_shift = 0, side = :last)
+    ub = length(tvec) + ub_shift
+    return if side == :last
+        clamp(searchsortedlastcorrelated(tvec, t, iguess) + idx_shift, lb, ub)
+    else
```
Comment (sathvikbhagavan) [inline src/interpolation_utils.jl:72]:
```suggestion
    elseif side == :first
```

File: src/interpolation_utils.jl [hunk 4]
```diff
@@ -54,11 +55,21 @@ function munge_data(U::StridedMatrix, t::AbstractVector)
...
+        clamp(searchsortedfirstcorrelated(tvec, t, iguess) + idx_shift, lb, ub)
+    end
```
Comment (sathvikbhagavan) [inline src/interpolation_utils.jl:74]: And an else condition to `error(...)`. This is to ensure it goes through the right code path.

**Comment by sathvikbhagavan** [comment]
@ChrisRackauckas, can you approve CI?

**Review by sathvikbhagavan — APPROVED** [review]

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**